### ISSUE TITLE
Styling the Focused DockWidget

### ISSUE THREAD

**Issue by dail8859** [githubuser0xFFFF/Qt-Advanced-Docking-System#163]
Using the examples in this repo, I've been able to style the CDockWidgetTab based on whether they are active or not, by the `activeTab` property. For example:

```css
ads--CDockWidgetTab[activeTab="false"]
{
	background: rgb(81, 81, 81);
	border-right: 1px solid rgb(51, 51, 51);
}
```

However I have not been able to determine a way style the *focused* CDockWidget differently. For a single DockManager there should be a max of 1 (or possibly 0) focused widgets. As real world examples, many editors have this, such as Visual Studio, VS Code, Notepad++ and others are able to do this. Here is an example from Visual Studio:

![image](https://user-images.githubusercontent.com/3694843/81486854-cf71ac00-9225-11ea-8422-46010bde060a.png)

For the 3 docked areas, each one has an "active" tab... `TextFile1.txt`, `TextFile3.txt`, and `TextFile6.txt` respectively. However it can be clearly seen that `TextFile3.txt` has the input focus. I don't believe this library is able to do this currently.

If the user clicks in a different part of the application, it can be seen that `TextFile3.txt` loses the focus:

![image](https://user-images.githubusercontent.com/3694843/81486905-4ad35d80-9226-11ea-81dc-720d1c062082.png)

While still being the "active" tab in that area.

---

I think having this ability for this library would be a great addition. I am willing to attempt the code changes needed as this would be incredibly beneficial for my use and I'm sure others. I'm not sure the best way to implement this functionality. I tried looking into using [`QApplication::focusChanged()`](https://doc.qt.io/qt-5/qapplication.html#focusChanged) signal which seems to be a good way to track what has the application's focus.

**Comment by githubuser0xFFFF** [comment]
Ok, I think, I understand what you mean. What would be the benefit for the user? We use this in a commercial product, an this is feature no user ever missed.

**Comment by dail8859** [comment]
In cases where multiple dockwidgets look similar it is beneficial to the user to see which one they are interacting with, especially when clicking toolbar buttons, menu items etc. 

I think a big difference is my use case involves alot of keyboard interaction (such as typing documents). Whereas other applications that are mostly mouse event based it is much easier for the user to be cognizant of which dockeidget they are interacting with since they click it often.

**Comment by githubuser0xFFFF** [comment]
I implemented this feature in the branch [focused_dockwidget](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/tree/focused_dockwidget). Please check it out, rebuild everything and test if it is what you want.

![focused_dockwidget](https://user-images.githubusercontent.com/4239304/81506326-cb14c400-92f5-11ea-8f39-0e6a4fd4be03.png)

**Comment by dail8859** [comment]
Thank you so much for the time and support! From the screenshot and glancing through the code changes it looks to be exactly what I was needing.

I will try these changes when I get a bit of free time and will report back.

**Comment by dail8859** [comment]
Sorry didn't mean to close this yet.

**Comment by dail8859** [comment]
@githubuser0xFFFF This seems to be exactly what I needed!

I did notice one small artifact. When a different tab is clicked then the focus is not forced to the DockWidget which does not style the tab as "focused" until the user clicks inside the DockWidget. Not sure if this is intentional...however even if the current behavior is kept I think this is fine as the application can easily catch and enforce its own focused state which will force the tab to be styled correctly.

Again I dont see this as a huge problem, but maybe worth documenting as to not cause confusion.

![activetab](https://user-images.githubusercontent.com/3694843/81507186-8f640500-92c9-11ea-8444-71d268ed4530.gif)

I'm not sure if it should be enforced or not, but honestly it is easy to do that on the application side

**Comment by githubuser0xFFFF** [comment]
You wrote 

> I have not been able to determine a way style the focused CDockWidget differently.

The focused CDockWidget is the one that contains a QWidget that has the focus. Switching the active dock widget via the tab bar does not switch the focus of any widget. So it is implemented the way you wrote.

If you would like the have a differend behaviour, I would be great if you could describe exactly what you would like to see.

**Comment by dail8859** [comment]
Visual Studio is probably the best example. Notice that clicking on a tab forces focus into the text editor (without having to click inside the editor, which allows a user to start typing right away) and also colors the tab the orangish color indicating it has the focus.

![vs](https://user-images.githubusercontent.com/3694843/81508036-82e2ab00-92cf-11ea-99b9-c64f0306b5a2.gif)

**Comment by githubuser0xFFFF** [comment]
Ok, I improved setting the focused dock widget tab a little bit. Now it is a little bit closer to VS. 

Something that is not working yet is updating the dock area style if the focused dock widget is closed.

**Comment by dail8859** [comment]
> Now it is a little bit closer to VS.

Yes definitely better! Very much appreciated.

> Something that is not working yet is updating the dock area style if the focused dock widget is closed.

Yes I see this behavior as well.

---

I found one other behavior. In my own application I programmatically create new DockWidgets and when new ones are added, it shows a different tab as active. I was able to reproduce this with the Demo app:

![newtabdemo](https://user-images.githubusercontent.com/3694843/81625644-7f324f80-93c7-11ea-8c9b-16c98769a3fc.gif)

Notice every time the `+` is clicked that it does not color the new tab with blue (focused).

---

Please don't take this as just me being nit-picky. I would certainly understand if this feature request begins to get too complex trying to handle every possible case. As with most software, it is easy to handle the nominal cases, and incredibly hard to handle all cases.

**Comment by KerimMatlab** [comment]
> I found one other behavior. In my own application I programmatically create new DockWidgets and when new ones are added, it shows a different tab as active. I was able to reproduce this with the

By the way the same thing happens in case of DockWidgets in Tabs when you change the active window via small popup triangle at the up-right corner (in the picture)
![1](https://user-images.githubusercontent.com/43808863/81626645-fcc88080-9404-11ea-8703-6a839a011f89.png)

**Comment by githubuser0xFFFF** [comment]
Thank you for testing. I hope I will have some time to improve this.

**Comment by githubuser0xFFFF** [comment]
I just pushed some commits, that should fix the issues mentioned above. Please test.

**Comment by KerimMatlab** [comment]
Thank you
Now it works fine except one special case.

I use VTK library. To visualize data inside Qt it I use [QVTKOpenGLStereoWidget](https://vtk.org/doc/nightly/html/classQVTKOpenGLWidget.html). Here is how I create this widget:
```
QVTKOpenGLStereoWidget* MainWindowObjectCreator::createVTK2DWindow(){
    QVTKOpenGLStereoWidget* qvtkOpenGLStereoWidget = new QVTKOpenGLStereoWidget;
    qvtkOpenGLStereoWidget->setFocusPolicy(Qt::StrongFocus);
    ads::CDockWidget* dockWidget = new ads::CDockWidget("2D Window");
    dockWidget->setWidget(qvtkOpenGLStereoWidget);
    dockManager->addDockWidget(ads::RightDockWidgetArea, dockWidget);
    return qvtkOpenGLStereoWidget;
}
```
When I launch the app I can't change focused windows by clicking on this `qvtkOpenGLStereoWidget` (see picture). But with all other widgets it works fine.
I understand that this is hardly connected with ADS but maybe you could make some assumption why it happens?
![1](https://user-images.githubusercontent.com/43808863/81811133-cede5b80-952c-11ea-96c4-10baad861f04.png)

**Comment by rolivav** [comment]
I fixed an issue I had to see if the focus enters or leaves a parent using `QApplication::focusChanged`. It could be used here too, I wrapped it in a class to reuse it:
https://gist.github.com/rolivav/ecc1c86b6e58ab4e204fc0c4f7e76782

**Comment by githubuser0xFFFF** [comment]
@KerimMatlab ADS uses `QApplication::focusChanged` signal to react on focus changes. So I think `qvtkOpenGLStereoWidget` does not support setting focus. Try to set the [focus policy](https://doc.qt.io/qt-5/qwidget.html#focusPolicy-prop) of this widget.

@rolivav Nice implementation  - Thank you

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF Yes I've tried to set focus policy to this VTK widget andthe following code can be compiled:
```
QVTKOpenGLStereoWidget* MainWindowObjectCreator::createVTK2DWindow(){
    QVTKOpenGLStereoWidget* qvtkOpenGLStereoWidget = new QVTKOpenGLStereoWidget;
    qvtkOpenGLStereoWidget->setFocusPolicy(Qt::ClickFocus);
    ads::CDockWidget* dockWidget = new ads::CDockWidget("2D Window");
    dockWidget->setWidget(qvtkOpenGLStereoWidget);
    dockManager->addDockWidget(ads::RightDockWidgetArea, dockWidget);
    return qvtkOpenGLStereoWidget;
}
```
But still clicking on this widget doesn't change the focused window

**Comment by githubuser0xFFFF** [comment]
Then this widget does not support focus and you need to implement this manually.

**Comment by KerimMatlab** [comment]
ok, thank you!

**Comment by dail8859** [comment]
@githubuser0xFFFF Thank you so much for continuing to refine this and glad to see others are interested as well.

Using the most recent commit (3a0c2a3113947dfc5e909e6f5b14c80205e6727c) only one other thing I noticed. If a new DockArea is created, it does not focus the newly dropped DockWidget.

To reproduce this:
1. Start the "sidebar" example
1. Merge the two DockWidgets into a single DockArea
1. Split the DockArea
1. Notice that the dropped DockWidget does not get styled as focused


--- 
![dropfocus](https://user-images.githubusercontent.com/3694843/81879527-0a8f1a80-9559-11ea-81f7-55a2607cec87.gif)

**Comment by githubuser0xFFFF** [comment]
Just pushed a fix that shuld improve this.

**Comment by dail8859** [comment]
The behavior seems to work much better now.

I've had a couple instances using the "sidebar" example, where when creating a new DockArea the dropped DockWidget is not marked as focused...however I have not been able to track down an exact set up steps to reproduce it yet. Once I do I will provide more information.

**Comment by githubuser0xFFFF** [comment]
Just pushed a commit to improve this further - now it should be almost ready for release. Please try, if you have still issues.

**Comment by dail8859** [comment]
Apologies on the long delay on this. I will try my best to test out the latest changes sometime today.

**Comment by dail8859** [comment]
I tried the current branch (819f1effc5a61047d330c8029990c83efe207a84) but it doesn't seem to be compiling on Windows for me. I had to comment out all the code inside of `updateFloatingWidgetFocusStyle()` so _I am not sure if that is the cause of what I am seeing_.

1. Run the deleteonclose example.
1. Create 3 tabs
1. Drag the last tab out so it is floating
1. Dock it to the right side.
1. Notice that it is not styled correctly

![nohighlight](https://user-images.githubusercontent.com/3694843/82505901-65ca8b00-9acc-11ea-98c8-801928232939.gif)

Again this could be a cause of the code I had to comment out.

**Comment by KerimMatlab** [comment]
I struggling with `focusChanged(QWidget *old, QWidget *now)`
I need to set property to every dock widget (it doesn't matter wether it's tabbed or float) and when the focus changed I need to retrieve the property from focused window.
The problem is that depending on wether dock widget is tabbed or float `QWidget *now` and `QWidget *old` will be `ads::CDockWidgetTab` or `QScrollArea::dockWidgetScrollArea` respectively.
So I can't find a way where to set my property to the `ads::CDockWidgetTab` or somewhere else.
For example if there is no float widgets then this code should set the property to the widget wich will be the `QWidget *now` when focusis changed:
```
QVTKOpenGLStereoWidget* MainWindowObjectCreator::createVTK2DWindow(){
    QVTKOpenGLStereoWidget* qvtkOpenGLStereoWidget = new QVTKOpenGLStereoWidget;
    ads::CDockWidget* dockWidget = new ads::CDockWidget("2D Window");
    dockWidget->setWidget(qvtkOpenGLStereoWidget);
    dockManager->addDockWidget(ads::RightDockWidgetArea, dockWidget);
    dockWidget->tabWidget()->setProperty("IsVisualization", 4443); // here it is
    //dockWidget->dumpObjectTree();
    return qvtkOpenGLStereoWidget;
}
```

**Comment by githubuser0xFFFF** [comment]
@dail8859 Fixed this. Focus styling is now an configuration option. Please enable the configuration flag `CDockManager::FocusStyling` to enable the focus feature.

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF Ain't I mistaken that you also fixed the top level widget that is focused? Seems to me that now it works correct and this is alway the `ads::CDockAreaTabBar`. Thank you! I'm testing it now

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF can't understand how to get `ads::CDockAreaTabBar` when I create `ads::CDockWidget* dockWidget = new ads::CDockWidget("2D Window");`?
I'm trying to set property to the `ads::CDockAreaTabBar` and then get this property in `QApplication::focusChanged`

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF here is small report:
My task is to set property on each dock widget and when the focused widget is changed I need to retrieve this property.
I set the property via: (I do this for two widgets and the second has a property value equal to 123)
```
QVTKOpenGLStereoWidget* MainWindowObjectCreator::createVTK2DWindow(){
    QVTKOpenGLStereoWidget* qvtkOpenGLStereoWidget = new QVTKOpenGLStereoWidget;
    ads::CDockWidget* dockWidget = new ads::CDockWidget("2D Window");
    dockWidget->setWidget(qvtkOpenGLStereoWidget);
    dockManager->addDockWidget(ads::RightDockWidgetArea, dockWidget);
    dockWidget->dockAreaWidget()->setProperty("IsVisualization", 4443);
    return qvtkOpenGLStereoWidget;
}
```
I retrieve the property with the command:
```
void MainWindow::onFocusChanged(QWidget* old, QWidget* now){
    if (now != nullptr){
        now->dumpObjectTree(); // just to see the structure of focused window
        ads::CDockAreaWidget* dockWidget = nullptr;
        while (now != nullptr && (dockWidget = qobject_cast<ads::CDockAreaWidget*>(now)) == nullptr)
            now = now->parentWidget();

        QVariant varia = dockWidget->property("IsVisualization");
    }
}
```
This works only if widgets are docked or float. 
But when I collect two widgets in **tab** then these two tabbed widgets has property wich alway gives me **123** no matter which tabbed widged is focused.
If then I make a float widget from one of these two tabbed widgets then I tgis float widget returns me unavailable property i.e the property doesn't work anymore...

**Comment by githubuser0xFFFF** [comment]
@KerimMatlab @dail8859 I just pushed some changes to this branche. The focus handling should be now in its final shape. There is now a `focusedDockWidgetChanged(CDockWidget* old, CDockWidget* now)` signal that enables the application to handle focus change. This is how the focussing it works:

- focus changes if a new tab is activated in a dock area (e.g. by mouse click, by tabs menu or if a noch dock area is added)
- focus changes if a widget in a dock widget gets focus
- right after start or after restoring a state, no widget is focused until the user exlicitely focusses a widget

Please test, if this works properly for you. If everything is o.k. for you, I'm going to merge the code into master

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF thank you!
I'm trying to build it with CMake and right after I press "Concfigure" I get the error:
**CMake Error at CMakeLists.txt:9 (project):
  VERSION "HEAD-HASH-NOTFOUND.HEAD-HASH-NOTFOUND.HEAD-HASH-NOTFOUND" format
  invalid.**
Do you have any idea why I can't build it now? Earlier there were not such problems. 
One possible reason I see is that few days ago I set some settings to my Git profile.
![image](https://user-images.githubusercontent.com/43808863/83869220-edeb9a00-a734-11ea-87e2-724498cb2715.png)

**Comment by githubuser0xFFFF** [comment]
Sorry, I have no idea. Usally I use qmake for the build. On the build server the CMake build seems to work. Maybe you should try a fresh checkout?

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF I built it with QMake and it works.
I tried to implement `focusedDockWidgetChanged(CDockWidget* old, CDockWidget* now)` but it seems doesn't work.
Here in class `MainWindowObjectCreator` I create `CDockManager` :
```
MainWindowObjectCreator::MainWindowObjectCreator(QMainWindow *parent)
    : QObject(parent)
{
    dockManager = new ads::CDockManager(parent);
    dockManager->setConfigFlag(ads::CDockManager::FocusStyling, true);
}
```
I connect the signal in `MainWindow` class:
```
MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    /* Focus chaged for QTreeWidget */
    connect(objCreator->dockManager, &ads::CDockManager::focusedDockWidgetChanged, this, &MainWindow::onFocusChanged);
}
```
and here is the implementation of slot in `MainWindow` class:
```
void MainWindow::onFocusChanged(ads::CDockWidget* old, ads::CDockWidget* now){
    if (now != nullptr){
        now->dumpObjectTree();
    }
}
```
`onFocusChanged` never works even I put a break point to anywhere in there.
Maybe beacause the focused window should be highlighted by blue color (in my app blue highlighting stopped to works I don't know why)?
![image](https://user-images.githubusercontent.com/43808863/83877716-b1736a80-a743-11ea-875c-80a2ecd9cdb3.png)

**Comment by githubuser0xFFFF** [comment]
Did you enable the `CDockmanager::FocusStyling` flag? Focus styling is disabled by default. You need to enable it explictely.

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF yes, as I showed above I wrote:
```
MainWindowObjectCreator::MainWindowObjectCreator(QMainWindow *parent)
    : QObject(parent)
{
    dockManager = new ads::CDockManager(parent);
    dockManager->setConfigFlag(ads::CDockManager::FocusStyling, true); // here I try to enable focusing
}
```
Ok, I will make small example and try to share it. Maybe there it is my fault in my code

**Comment by githubuser0xFFFF** [comment]
Does the demo application work for you?

**Comment by githubuser0xFFFF** [comment]
Btw, you need to set the config flag before you create the dock manager instance like it is done in the demo application.

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF oh thank you! now it works.
I'm going to test it today (probably now) and will report soon.

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF It seems that it work almost perfectly. No matter wether I click on on button which belongs to some `CDockWidget` or to some space of other `CDockWIdget` focus is changed. I like this, thank you!

The only thing is about `QVTKOpenGLStereoWidget` wich [we discussed above](https://github.com/githubuser0xFFFF/Qt-Advanced-Docking-System/issues/163#issuecomment-628215598) and when I click on this widget the focused `CDockWidget` doesn't change.
So when I hange focused `CDockWidget` I get application output like:

> CDockManager::onFocusChanged
> FocuseNow ads::CDockWidgetTab
> CDockManager::onFocusChanged Horizons
> !!!!!!!!!!!! focusedDockWidgetChanged Read SEGY -> Horizons
> FocusedArea
> CDockManager::onFocusChanged

But when I click right on the `QVTKOpenGLStereoWidget` I get:

> CDockManager::onFocusChanged

but parent of this `CDockWidget` doesn't become blue (focused). Maybe you have an idea how to deal with that.

But anyway, thank you!

**Comment by dail8859** [comment]
@githubuser0xFFFF These changes sound exactly what I'd need. I've been incredibly busy lately so I will try to make some time this weekend to try out these changes. Thanks again for all your effort!

**Comment by githubuser0xFFFF** [comment]
@KerimMatlab I just pushed a new commit that enabled `Qt::ClickFocus` for `CDockWidget`. Please try, if this fixes your focus issue with `QVTKOpenGLStereoWidget`.

Please note that I renamed the flag `CDockManager::FocusStyling` into `CDockManager::FocusHighlighting`.

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF I just tried to recompile `ads` with `FocusHighlighting` but it still doesn't work with `QVTKOpenGLStereoWidget`.
Normal output is:

> CDockManager::onFocusChanged
> focusedNow 0000017B68673020
> FocuseNow QTreeView
> CDockManager::onFocusChanged Wells

but when I click on `QVTKOpenGLStereoWidget` I get:

> CDockManager::onFocusChanged
> focusedNow 0000000000000000

and the window doesn't become blue.

**Comment by githubuser0xFFFF** [comment]
@KerimMatlab ADS can only detect focus changed, if your DockWidget content supports focussing. I just tested it with `QVTKOpenGLStereoWidget` and can see the same that you see - the focused dock widget is a nulltptr if I click into this widget. If you have a widget that does not properly support focus, then you need to implement it yourself.

I could properly add focus highlighting for this widget by installing an event filter on it and setting the focus on `QEvent::MouseButtonPress` event. This is the function for creation of the vtk Widget:

```c++
//============================================================================
static ads::CDockWidget* createVTK2DWindow(QMenu* ViewMenu, QObject* EventFilter)
{
    QVTKOpenGLStereoWidget* qvtkOpenGLStereoWidget = new QVTKOpenGLStereoWidget;
    ads::CDockWidget* DockWidget = new ads::CDockWidget("2D Window");
    DockWidget->setWidget(qvtkOpenGLStereoWidget);
    ViewMenu->addAction(DockWidget->toggleViewAction());
    qvtkOpenGLStereoWidget->installEventFilter(EventFilter);
    qvtkOpenGLStereoWidget->setProperty("DockWidget", QVariant::fromValue(DockWidget));
    return DockWidget;
}
```

And this is the event filter function:

```c++
bool CMainWindow::eventFilter(QObject *watched, QEvent *event)
{
    qDebug() << event;
    if (event->type() == QEvent::MouseButtonPress)
    {
        qDebug() << "watched: " << watched;
        QVTKOpenGLStereoWidget* vtkWidget =  qobject_cast<QVTKOpenGLStereoWidget*>(watched);
        auto vDockWidget = vtkWidget->property("DockWidget");
        ads::CDockWidget* DockWidget = nullptr;
        if (vDockWidget.isValid())
        {
            DockWidget = qvariant_cast<ads::CDockWidget*>(vDockWidget);
        }

        if (DockWidget)
        {
            d->DockManager->setDockWidgetFocused(DockWidget);
        }
    }
    return false;
}
```

So, to summarize: either your widget properly supports focussing or does not block focussing of the CDockWidget or you need to implement this by yourself.

**Comment by dail8859** [comment]
@githubuser0xFFFF 

I just tried the most recent commit (a9268e6bf70f48a41989abe5223ba78496fdca47) and during my brief testing everything is working perfectly for me and behaves exactly as expected so far.

Thank you for the effort to get this working. I am very glad that this functionality is getting added...and it appears others will find this useful as well.

👍

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF thank you for cemprehensive answer and help
I think `eventFilter` would be enough for me.
Sorry for answer delay

**Comment by githubuser0xFFFF** [comment]
Just mergef this into master

**Comment by KerimMatlab** [comment]
@githubuser0xFFFF excuse me, would it be difficult to if I ask reimplement `CDockManager::FocusHighlighting` to `CDockWidget::FocusHighlighting`?
That would allow to some `CDockWidget` be focused and other `CDockWidget` must not be focused.
I'm really sorry that I noticed that thing only now after you already merged branches

**Comment by derekvantonder** [comment]
> @githubuser0xFFFF thank you! I'm trying to build it with CMake and right after I press "Concfigure" I get the error: **CMake Error at CMakeLists.txt:9 (project): VERSION "HEAD-HASH-NOTFOUND.HEAD-HASH-NOTFOUND.HEAD-HASH-NOTFOUND" format invalid.** Do you have any idea why I can't build it now? Earlier there were not such problems. One possible reason I see is that few days ago I set some settings to my Git profile. ![image](https://user-images.githubusercontent.com/43808863/83869220-edeb9a00-a734-11ea-87e2-724498cb2715.png)

I ran into this today. The solution I found that worked for me was to CLONE THE GITHUB REPO for ADS, and then configure/generate CMake from there using the 'Unix Makefiles' option (I'm on Mac OS).

I got this error when I had just downloaded a ZIP file containing the source and tried to compile it, but it doesn't seem to like that (might be something for the creators to look into?). 

I noted that it was requesting all kinds of Git data where it ran into this error, so presumably it only works if you are using a CLONED REPO from Git, instead of just an isolated directory copied from a ZIP.

Now it works great, wonderful work @githubuser0xFFFF! What a fantastic library.

**Comment by hpdumm** [comment]
I was able to build from the tar.gz file by passing "-DADS_VERSION:STRING=4.2.1" to cmake. This resolved the 'VERSION "HEAD-HASH-NOTFOUND.HEAD-HASH-NOTFOUND.HEAD-HASH-NOTFOUND" format invalid' error.
